### PR TITLE
DDOG Integration

### PR DESCRIPTION
**Pull request by elijahbenizzy**
Iterating -- plan of attack/thoughts:

- [x] Opted to store a cache myself and prune rather than using context as I want to leave the door open for more clever uses of this. I just link parents and store a dict
- [x] Should be easy to get tasks working, just need to expose the API
- [x] Should also be possible to add causal links between spans as well. TBD on how to use the DDOG API exactly.
- [x] Add all node tags to DDOG cause why not :shrug:
- [ ] do OpenTel (out of scope for this, that's next)

What this looks like:

image

## Changes
- DDOG
- An extra (easter-egg) hook
- Improvements to running hooks in the case of exceptions

## How I tested this
- ran ddog multiple times locally under different contexts
- tests passed for existing ones (what I was fixing for was not worth testing for TBH, parity with current behavior is good)

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: hamilton/lifecycle/base.py [hunk 1]
```diff
@@ -751,19 +751,21 @@ def _get_lifecycle_validators(
...
-        async_hooks = collections.defaultdict(set)
+        sync_hooks = collections.defaultdict(list)
+        async_hooks = collections.defaultdict(list)
         for adapter in self.adapters:
             for cls in inspect.getmro(adapter.__class__):
                 sync_hook = getattr(cls, SYNC_HOOK, None)
                 if sync_hook is not None:
-                    sync_hooks[sync_hook].add(adapter)
+                    if adapter not in sync_hooks[sync_hook]:
+                        sync_hooks[sync_hook].append(adapter)
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/base.py:761]: OK, this is not ironed into the contract, and we'll likely want to switch around order, but I think for the future we need to think about this a bit -- my instinct says to do `pre_` in order and `post_` in reverse order, but that's complicated and ambiguous to implement.

File: hamilton/plugins/h_ddog.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
+from typing import Any, Dict, Optional
+
+from ddtrace import Span, context, tracer
+
+from hamilton import lifecycle
+
+
+class DDOGTracer(
```
Comment (elijahbenizzy) [inline hamilton/plugins/h_ddog.py:8]: Could add testing but I don't think its worth it, I'll just fix forward and if its a pain add testing

File: hamilton/lifecycle/default.py [hunk 3]
```diff
@@ -307,3 +308,33 @@ def run_after_node_execution(
...
+
+
+def wait_random(mean: float, stddev: float):
+    sleep_time = random.gauss(mu=mean, sigma=stddev)
+    if sleep_time < 0:
+        sleep_time = 0
+    time.sleep(sleep_time)
+
+
+class SlowDownYouMoveTooFast(NodeExecutionHook):
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/default.py:320]: Easter egg

File: hamilton/lifecycle/base.py (hunk 1 above)
Comment (skrawcz) [reply hamilton/lifecycle/base.py:761]: yes we should do it that way ideally

File: hamilton/plugins/h_ddog.py [hunk 4]
```diff
@@ -1,78 +1,192 @@
-from typing import Any, Dict, Optional, Tuple
+from typing import Any, Dict, Optional
 
-from ddtrace import tracer
+from ddtrace import Span, context, tracer
```
Comment (skrawcz) [inline hamilton/plugins/h_ddog.py]: surround with try/except and print what to install to use this ?

File: hamilton/plugins/h_ddog.py [hunk 5]
```diff
@@ -1,78 +1,192 @@
...
-        )  # Cache of (run_id, task_id, node_id) tuples -- carries multiple levels
+        )  # Cache of run_id -> [task_id, node_id] -> span. We use this to open/close general traces
 
     @staticmethod
-    def _span_key(
-        run_id: str, task_id: Optional[str], node_id: Optional[str]
-    ) -> Tuple[Optional[str], ...]:
-        return run_id, task_id, node_id
+    def _serialize_span_dict(span_dict: Dict[str, Span]):
+        """Serializes to a readable format. We're not propogating links, but that's fine (for now).
```
Comment (skrawcz) [inline hamilton/plugins/h_ddog.py:49]: what do you mean link?

**Review by skrawcz — APPROVED** [review]

File: hamilton/plugins/h_ddog.py (hunk 5 above)
Comment (elijahbenizzy) [reply hamilton/plugins/h_ddog.py:49]: Span links -- see notes elsewhere

**Pull request merged**